src/adapter/src/coord/introspection.rs

```diff
@@ -0,0 +1,199 @@
+        MZ_INTROSPECTION_ROLE, PG_CATALOG_SCHEMA,
+    },
+    AdapterError,
+};
+
+/// The schema's a user is allowed to query from the `mz_introspection` cluster
+static ALLOWED_SCHEMAS: Lazy<HashSet<&str>> = Lazy::new(|| {
+    HashSet::from([
+        MZ_CATALOG_SCHEMA,
+        PG_CATALOG_SCHEMA,
+        MZ_INTERNAL_SCHEMA,
+        INFORMATION_SCHEMA,
+    ])
+});
+
+/// Checks whether or not we should automatically run a query on the `mz_introspection`
+/// cluster, as opposed to whatever the current default cluster is.
+pub fn auto_run_on_introspection<'a, 's>(
+    catalog: &'a Catalog,
+    session: &'s mut Session,
+    depends_on: impl IntoIterator<Item = GlobalId>,
+) -> Option<&'a Cluster> {
+    // If this feature is disabled via LaunchDarkly, or the user has disabled it for
+    // this session.
+    if !catalog.system_config().enable_force_introspection_cluster()
+        || !session.vars().force_introspection_cluster()
+    {
+        return None;
+    }
+
+    // Check to make sure our iterator contains atleast one element, this prevents us
+    // from always running empty queries on the mz_introspection cluster.
+    let mut depends_on = depends_on.into_iter().peekable();
+    let non_empty = depends_on.peek().is_some();
+
+    // Make sure we only depend on the system catalog.
+    let system_only = depends_on.all(|id| {
+        let entry = catalog.get_entry(&id);
+        let full_name = catalog.resolve_full_name(entry.name(), Some(session.conn_id()));
+        ALLOWED_SCHEMAS.contains(full_name.schema.as_str())
+    });
+
+    // If we're allowed to run on the mz_introspection cluster, make sure we can resolve it.
+    if non_empty && system_only {
+        let intros_cluster = catalog.resolve_cluster(MZ_INTROSPECTION_CLUSTER.name);
+        let active_cluster = catalog.active_cluster(session);
+
+        match (intros_cluster, active_cluster) {
+            (Ok(intros), active) => {
+                tracing::debug!("Running on mz_introspection cluster");
+                // If we're running on a different cluster than the active
+                // one, notify the user.
+                if active.map(|c| c.id != intros.id).unwrap_or(true) {
+                    session.add_notice(AdapterNotice::AutoRunOnIntrospectionCluster);
+                }
+                Some(intros)
+            }
+            (Err(e), _) => {
+                // Uh oh, failed to resolve the mz_introspection cluster, nothing we can do.
+                tracing::error!("Failed to resolve mz_introspection cluster, fall {:?}", e);
+                None
+            }
+        }
+    } else {
+        None
+    }
+}
+
+/// TODO(jkosh44) This function will verify the privileges for the mz_introspection user.
+///  All of the privileges are hard coded into this function. In the future if we ever add
+///  a more robust privileges framework, then this function should be replaced with that
+///  framework.
+pub fn user_privilege_hack(
```

Comment:
This function is just a refactor/moving it from where it previously existed and into this new module.